exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -223,18 +223,29 @@ impl Dispatcher {
     }
 
     /// Calls `before_commit` for all currently active services, isolating each call.
-    pub(crate) fn before_commit(&self, fork: &mut Fork) {
-        for (&service_id, info) in &self.service_infos {
-            let context = ExecutionContext::new(self, fork, Caller::Blockchain);
-            if self.runtimes[&info.runtime_id]
-                .before_commit(context, service_id)
-                .is_ok()
-            {
-                fork.flush();
-            } else {
-                fork.rollback();
-            }
-        }
+    ///
+    /// # Return value
+    ///
+    /// Returns errors that occurred during the calls.
+    pub(crate) fn before_commit(&self, fork: &mut Fork) -> Vec<(InstanceId, ExecutionError)> {
+        self.service_infos
+            .iter()
+            .filter_map(|(&service_id, info)| {
+                let context = ExecutionContext::new(self, fork, Caller::Blockchain);
+                let result = self.runtimes[&info.runtime_id].before_commit(context, service_id);
+
+                match result {
+                    Ok(()) => {
+                        fork.flush();
+                        None
+                    }
+                    Err(e) => {
+                        fork.rollback();
```

Comment:
Offtopic: Hm, seems that there is implicit expectation for the callee to provide a flushed fork to this method. Maybe it makes sense to either document this as a pre-requirement, or do the fork flush before processing services.